rest_framework/utils/html.py

```diff
@@ -12,7 +12,7 @@ def is_html_input(dictionary):
     return hasattr(dictionary, 'getlist')
 
 
-def parse_html_list(dictionary, prefix=''):
+def parse_html_list(dictionary, prefix='', default=[]):
```

Comment:
Do we need this argument? At **all** call sites we're passing `empty`. (Maybe there are third-party users. This would maintain BC.)

Reply:
Two (and a half) reasons why I implemented it this way

### Reason 1: Compatibility

The way it used to be was this:
```python
def parse_html_list(dictionary, prefix=''):
    ret = {}
    # magic happens here :)
    return [ret[item] for item in sorted(ret)]
```

Now if the ``ret`` dict being empty, ``parse_html_list`` returns an empty list.

As you already said, the ``default=[]`` would help to avoid a breaking change for 3rd party libraries (and to be fair, ``parse_html_list``

### Reason 2 (and a half): Codestyle and DRY

The most obvious alternative would be this:

```python
def parse_html_list(dictionary, prefix=''):
    ret = {}
    # magic happens here :)
    from rest_framework.fields import empty
    return [ret[item] for item in sorted(ret)] if len(ret.keys()) > 0 else empty
```

However, we can (should) not ``import empty`` at the top of ``utils/html.py``, as it is a utility file, and it will lead to cyclic imports. Hence we would have to import it within the ``parse_html_list`` function. And that just looks wrong to me from multiple perspectives (code style, complexity, expectations on code).

The other (less obvious and less dry) alternative is to look at all call sites of ``parse_html_list`` and change the call from
```python
return html.parse_html_list(dictionary, prefix=self.field_name, default=empty)
```
to
```python
ret_val = html.parse_html_list(dictionary, prefix=self.field_name)
return ret_val if len(ret_val) == 0 else empty
```

We would have to do that at several places within the code, but it seems okay for me.


Also, I want to address your statement:
> At all call sites we're passing empty

Actually, I deliberately did not touch the ``ListField.to_internal_value`` method here:
https://github.com/anx-ckreuzberger/django-rest-framework/blob/3d51bcf54ea84e5030db01bb6dbd274abbc4d0bc/rest_framework/fields.py#L1621-L1631

I am unsure if returning ``empty`` for the ``ListField`` here would cause any issues or not. I only modified those calls to ``parse_html_list``, where to parent method returns ``empty`` in certain cases anyway, such as here:
https://github.com/anx-ckreuzberger/django-rest-framework/blob/3d51bcf54ea84e5030db01bb6dbd274abbc4d0bc/rest_framework/fields.py#L1606-L1619

Reply:
OK thanks @anx-ckreuzberger. That all make sense.

Reply:
You could pass `list` itself. Then:

```
if default is list:
    default = list()
```

This allows using `None` as an actual default.

Reply:
So we make ``default=None`` the default behaviour, and I'll overwrite with default=list for the function calls where appropriate?

Reply:
I'm not sure what `I'll overwrite with default=list for the function calls` means exactly. I don't think there's any reason we should be using the base `list` type here.

Reply:
Sorry, I think I didn't get my question across to you :) Would you rather want to see
```python
def parse_html_list(dictionary, prefix='', default=list):
    if default is list:
        default = list()
```
or 
def parse_html_list(dictionary, prefix='', default=None):
    if default is list:
        default = list()
```

Reply:
The second, `default=None` :)

Reply:
Neither. I don't think `if default is list` makes sense - why would we be passing the bare type?

Reply:
I was referring to the comment of @carltongibson to avoid having a mutable as a default parameter.

>You could pass list itself. Then:
```python
if default is list:
    default = list()
```
>This allows using None as an actual default.

But I'll just set the default to None, and discard the ``if default is list`` idea :)

Reply:
👍